vyper/parser/context.py

```diff
@@ -71,25 +72,35 @@ def __init__(self, vars, global_ctx, sigs=None, forvars=None, return_type=None,
         # store global context
         self.global_ctx = global_ctx
 
-    def set_in_assignment(self, state: bool):
-        self.in_assignment = state
+    def is_constant(self):
+        return self.constancy == Constancy.Constant or self.in_assertion
 
-    def set_in_for_loop(self, name_of_list):
-        self.in_for_loop.add(name_of_list)
+    #
+    # Context Managers
+    # - Context managers are used to ensure proper wrapping of scopes and context states.
 
-    def remove_in_for_loop(self, name_of_list):
+    @contextmanager
+    def set_in_for_loop_scope(self, name_of_list):
+        self.in_for_loop.add(name_of_list)
+        yield
         self.in_for_loop.remove(name_of_list)
 
-    def is_constant(self):
-        return self.constancy == Constancy.Constant or self.in_assertion
+    @contextmanager
+    def set_in_assignment(self):
+        self.in_assignment = True
+        yield
+        self.in_assignment = False
```

Comment:
I *think* the current pattern is right, but I want to point out that in many cases, you will want to structure this as:

```
try:
    yield
finally:
    self.in_assignment = False
```

Without this, the final cleanup step won't happen if an exception occurs within the context block.  I'm not sure which would be your desired behavior in this situation but it appears that the current code mirrors the previous functionality.

Reply:
Noted: Yes, I think for these cases we don't have sockets/files/dbs etc. to close so it's _soft_ with statement ;)